src/v/cloud_topics/reconciler/reconciler.cc

```diff
@@ -180,174 +232,353 @@ ss::future<> reconciler::reconciliation_loop() {
+        // TODO: The object has been uploaded. The reconciler could
+        //       attempt cleanup (or notify a cleanup subsystem).
+        co_return std::unexpected(reconcile_error::metadata_failure);
+    }
 
-    vlog(
-      lg.debug,
-      "Committed overlay to object {} for {} log {}~{}. New LRO {}",
-      object_id,
-      part->ntp(),
-      metadata.base_offset,
-      metadata.last_offset,
-      partition_info.partition->lro);
+    co_return std::expected<void, reconcile_error>{};
+}
 
-    co_return;
+ss::future<std::expected<void, reconcile_error>> reconciler::commit_objects(
+  const std::vector<object_metadata>& objects,
+  std::unique_ptr<l1::metastore::object_metadata_builder> meta_builder) {
+    // It's possible to build the terms map as we build the objects, but
+    // I think re-iterating over all the object partitions here is worth
+    // it in exchange for less context passing among functions.
+    l1::metastore::term_offset_map_t terms;
+    for (const auto& obj_meta : objects) {
+        for (const auto& partition : obj_meta.partitions) {
+            chunked_vector<l1::metastore::term_offset> term_offsets;
+            for (const auto& [term, first_offset] : partition.metadata.terms) {
+                term_offsets.emplace_back(term, first_offset);
+            }
+            terms[partition.partition->tidp] = std::move(term_offsets);
+        }
+    }
+
+    auto add_objects_result = co_await _metastore->add_objects(
+      std::move(meta_builder), terms);
+    if (!add_objects_result.has_value()) {
+        vlog(
+          lg.error,
+          "Failed to add objects to the L1 metastore: {}",
+          add_objects_result.error());
+        // TODO: The objects have been uploaded. The reconciler could
+        //       attempt cleanup (or notify a cleanup subsystem).
+        co_return std::unexpected(reconcile_error::metadata_failure);
+    }
+
+    // Now update the LRO, taking into account any corrections from
+    // the metastore.
+    // TODO: Inform L0 of the update.
+    const auto& corrected_next_offsets
+      = add_objects_result.value().corrected_next_offsets;
+    for (const auto& obj_meta : objects) {
+        for (const auto& partition : obj_meta.partitions) {
+            auto next_lro = kafka::offset::min();
+            if (corrected_next_offsets.contains(partition.partition->tidp)) {
+                next_lro = corrected_next_offsets.at(partition.partition->tidp);
+            } else {
+                auto [first, last] = obj_meta.object_info.index.partitions
+                                       .equal_range(partition.partition->tidp);
+                for (auto it = first; it != last; ++it) {
+                    const auto& obj_partition = it->second;
+                    next_lro = std::max(
+                      next_lro, kafka::next_offset(obj_partition.last_offset));
+                }
+            }
+            partition.partition->lro = next_lro;
```

Comment:
This state is not persisted anywhere, we should do something like this:

```cpp
    auto stm
      = range.partition->partition->raft()->stm_manager()->get<ctp_stm>();
    vassert(stm, "ctp_stm must be present on cloud topic partitions");

    retry_chain_node rtc(_as);
    ctp_stm_api api(rtc, stm);

    auto ret = co_await api.advance_reconciled_offset(
          next_lro);
    if (!ret.has_value()) {
          vlog(
            lg.info,
            "Failed to advance LRO to {}: {}",
            range.partition->lro,
            ret.error());
          co_return;
    }
```

Reply:
good catch. @wdberkeley did you have different/later plans for that?

Reply:
https://redpandadata.atlassian.net/browse/CORE-13424